From ebcb99c1a3d35055b07c0a02857ffab3226ba9e5 Mon Sep 17 00:00:00 2001
From: Stefan Roese <sr@denx.de>
Date: Mon, 6 May 2019 15:09:51 +0200
Subject: [PATCH] i2c: mt7621: Configure controller pins to open-drain instead
 of push-pull

This patch changes the configuration of the I2C pins to use open-drain
instead of push-pull. This is needed for correct I2C communication. The
problem results from a misleading bit description for the open-drain
setup (inverted meaning).

Using push-pull might lead to issues on complexer I2C infrastructures
and / or higher I2C bus speeds.

Signed-off-by: Stefan Roese <sr@denx.de>
---
 drivers/i2c/busses/i2c-mt7621.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/i2c/busses/i2c-mt7621.c b/drivers/i2c/busses/i2c-mt7621.c
index 734b6f5434b5..c2d5b212d12f 100644
--- a/drivers/i2c/busses/i2c-mt7621.c
+++ b/drivers/i2c/busses/i2c-mt7621.c
@@ -128,11 +128,12 @@ static void mtk_i2c_reset(struct mtk_i2c *i2c)
 {
 	device_reset(i2c->adap.dev.parent);
 	barrier();
-	mtk_i2c_w32(i2c,
-		   SM0CTL0_ODRAIN
-		   | ((i2c->clk_div << 16) & SM0CTL0_CLK_DIV_MASK)
-		   | SM0CTL0_EN
-		   | SM0CTL0_SCL_STRETCH, REG_SM0CTL0_REG);
+	/*
+	 * Don't set SM0CTL0_ODRAIN as its bit meaning is inverted. To
+	 * configure open-drain mode, this bit needs to be cleared.
+	 */
+	mtk_i2c_w32(i2c, ((i2c->clk_div << 16) & SM0CTL0_CLK_DIV_MASK) |
+		    SM0CTL0_EN | SM0CTL0_SCL_STRETCH, REG_SM0CTL0_REG);
 	mtk_i2c_w32(i2c, 0, REG_SM0CFG2_REG);
 }
 
